### PR TITLE
Fix reflector List cancellation

### PR DESCRIPTION
**Pull request by tnozicka**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
When closing stopCh informer could hang on List call indefinitely.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/cc @kubernetes/sig-api-machinery-bugs

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 1]
```diff
@@ -175,10 +175,31 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 	options := metav1.ListOptions{ResourceVersion: "0"}
 	r.metrics.numberOfLists.Inc()
 	start := r.clock.Now()
-	list, err := r.listerWatcher.List(options)
+
+	var list runtime.Object
+	var err error
+	listCh := make(chan struct{}, 1)
```
Comment (sttts) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go]: why 1? why buffer at all?

Comment (sttts) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go]: make it a `chan err` and get rid of the shared variable, same for panic.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 2]
```diff
@@ -175,10 +175,31 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
...
+			}
+		}()
+		list, err = r.listerWatcher.List(options)
+		close(listCh)
+	}()
+	select {
+	case <-stopCh:
+		return nil
+	case r := <-panicCh:
+		panic(r)
```
Comment (sttts) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go]: note: this way you lose the backtrace of the actual issue.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 1 above)
Comment (tnozicka) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: if you get stopped the goroutine would hang blocking on the channel write

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: it's not easy to change that though. In the apiserver handler code we append errors with the backtrace string. Not nice either. Maybe just leave it here as it is.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 1 above)
Comment (tnozicka) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: it returns runtime.object as well making you wrap stuff in structs for the tuple which didn't seem to worth it

**Pull request merged**